File: src/analyzers/cyclomatic.ts
import type { Analyzer, AnalyzerResult, AnalyzerDetail } from './types.js';
import type { CodeBenchConfig } from '../config/index.js';
import type { ParsedFile } from '../parsers/index.js';

/**
 * Count cyclomatic complexity for a function body.
 * Each decision point adds 1: if, else if, for, while, do, case, catch, &&, ||, ??, ternary.
 * Base complexity is 1.
 */
function computeCyclomaticComplexity(body: string): number {
  let complexity = 1;

  const decisionPatterns = [
    /\bif\s*\(/g,
    /\belse\s+if\s*\(/g,
    /\bfor\s*\(/g,
    /\bwhile\s*\(/g,
    /\bdo\s*\{/g,
    /\bcase\s+/g,
    /\bcatch\s*\(/g,
    /&&/g,
    /\|\|/g,
    /\?\?/g,
    /\?[^?:]*:/g, // ternary (rough)
  ];

  for (const pattern of decisionPatterns) {
    const matches = body.match(pattern);
    if (matches) complexity += matches.length;
  }

  // Subtract double-counted else-if (counted in both if and else if)
  const elseIfMatches = body.match(/\belse\s+if\s*\(/g);
  if (elseIfMatches) complexity -= elseIfMatches.length;

  return complexity;
}

export const cyclomaticAnalyzer: Analyzer = {
  name: 'cyclomatic',

  async analyze(
    files: ParsedFile[],
    config: CodeBenchConfig,
  ): Promise<AnalyzerResult> {
    const threshold = config.thresholds.max_cyclomatic;
    const details: AnalyzerDetail[] = [];
    let totalFunctions = 0;
    let functionsWithinThreshold = 0;

    for (const file of files) {
      for (const fn of file.functions) {
        const complexity = computeCyclomaticComplexity(fn.body);
        totalFunctions++;

        if (complexity <= threshold) {
          functionsWithinThreshold++;
        } else {
          details.push({
            file: file.relativePath,
            message: `${fn.name}() has cyclomatic complexity ${complexity} (threshold: ${threshold})`,
            line: fn.startLine,
            value: complexity,
          });
        }
      }
    }

    if (totalFunctions === 0) {
      return {
        metric: 'cyclomatic',
        score: 100,
        details: [],
        summary: 'No functions found to analyze',
      };
    }

    const score = Math.round((functionsWithinThreshold / totalFunctions) * 100);
    const violations = totalFunctions - functionsWithinThreshold;

    return {
      metric: 'cyclomatic',
      score,
      details: details.sort((a, b) => (b.value ?? 0) - (a.value ?? 0)),
      summary: `${violations} of ${totalFunctions} functions exceed complexity threshold of ${threshold}`,
    };
  },
};

File: internal/analyzer/cyclomatic.go
package analyzer

import (
	"fmt"
	"math"
	"regexp"
	"sort"

	"github.com/codebench/codebench/internal/config"
	"github.com/codebench/codebench/internal/parser"
)

type CyclomaticAnalyzer struct{}

var cyclomaticPatterns = []*regexp.Regexp{
	regexp.MustCompile(`\bif\s*\(`),
	regexp.MustCompile(`\belse\s+if\s*\(`),
	regexp.MustCompile(`\bfor\s*\(`),
	regexp.MustCompile(`\bwhile\s*\(`),
	regexp.MustCompile(`\bdo\s*\{`),
	regexp.MustCompile(`\bcase\s+`),
	regexp.MustCompile(`\bcatch\s*\(`),
	regexp.MustCompile(`&&`),
	regexp.MustCompile(`\|\|`),
	regexp.MustCompile(`\?\?`),
	regexp.MustCompile(`\?[^?:]*:`),
}

var elseIfPattern = regexp.MustCompile(`\belse\s+if\s*\(`)

func computeCyclomaticComplexity(body string) int {
	complexity := 1

	for _, pattern := range cyclomaticPatterns {
		matches := pattern.FindAllString(body, -1)
		complexity += len(matches)
	}

	// Subtract double-counted else-if
	elseIfMatches := elseIfPattern.FindAllString(body, -1)
	complexity -= len(elseIfMatches)

	return complexity
}

func (a *CyclomaticAnalyzer) Name() config.MetricName {
	return config.MetricCyclomatic
}

func (a *CyclomaticAnalyzer) Analyze(files []*parser.ParsedFile, cfg *config.Config, cwd string) (*Result, error) {
	threshold := cfg.Thresholds.MaxCyclomatic
	var details []Detail
	totalFunctions := 0
	withinThreshold := 0

	for _, file := range files {
		for _, fn := range file.Functions {
			complexity := computeCyclomaticComplexity(fn.Body)
			totalFunctions++

			if complexity <= threshold {
				withinThreshold++
			} else {
				details = append(details, Detail{
					File:    file.RelativePath,
					Message: fmt.Sprintf("%s() has cyclomatic complexity %d (threshold: %d)", fn.Name, complexity, threshold),
					Line:    fn.StartLine,
					Value:   float64(complexity),
				})
			}
		}
	}

	if totalFunctions == 0 {
		return &Result{
			Metric:  config.MetricCyclomatic,
			Score:   100,
			Summary: "No functions found to analyze",
		}, nil
	}

	score := int(math.Round(float64(withinThreshold) / float64(totalFunctions) * 100))
	violations := totalFunctions - withinThreshold

	sort.Slice(details, func(i, j int) bool {
		return details[i].Value > details[j].Value
	})

	return &Result{
		Metric:  config.MetricCyclomatic,
		Score:   score,
		Details: details,
		Summary: fmt.Sprintf("%d of %d functions exceed complexity threshold of %d", violations, totalFunctions, threshold),
	}, nil
}
